OpenRA.Mods.Common/Traits/GrantConditionWhenDockClient.cs

```diff
@@ -0,0 +1,82 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright (c) The OpenRA Developers and Contributors
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public sealed class GrantConditionWhenDockClientInfo : TraitInfo
+	{
+		[FieldLoader.Require]
+		[GrantedConditionReference]
+		[Desc("The condition to grant to self")]
+		public readonly string Condition = null;
+
+		[Desc("How long condition is applid even after undock. Use -1 for infinite.")]
+		public readonly int AfterDockDuration = 0;
+
+		[Desc("Host actor leading to the condition being granted. Enable for all at default.")]
+		public readonly HashSet<string> DockHostNames = null;
+
+		public override object Create(ActorInitializer init) { return new GrantConditionWhenDockClient(this); }
+	}
+
+	public sealed class GrantConditionWhenDockClient : INotifyDockClient, ITick, ISync
+	{
+		readonly GrantConditionWhenDockClientInfo info;
+		int token;
+		int delayedtoken;
+
+		[Sync]
+		public int Duration { get; private set; }
+
+		public GrantConditionWhenDockClient(GrantConditionWhenDockClientInfo info)
+		{
+			this.info = info;
+			token = Actor.InvalidConditionToken;
+			delayedtoken = Actor.InvalidConditionToken;
+		}
+
+		void INotifyDockClient.Docked(Actor self, Actor host)
+		{
+			if (info.Condition != null && (info.DockHostNames == null || info.DockHostNames.Contains(host.Info.Name)))
+			{
+				if (token == Actor.InvalidConditionToken)
```

Comment:
Doesn't this also need to check if `delayedToken` is currently set? Otherwise we will grant the condition twice! I suggest setting `token` to `delayedToken` (when `delayedToken != Actor.InvalidContionToken`) and resetting `delayedToken` to `Actor.InvalidContionToken`.

Reply:
Wow, thanks! I missed this fatal problem!

Reply:
Fixed